File: frontend/src/stores/appAuth.ts
import { computed, ref } from 'vue'
import { defineStore } from 'pinia'
import { appAuthApi } from '@/api'
import type {
  AppAuthStatus,
  AppLoginRequest,
  AppPasswordChangeRequest,
  AppSetupRequest,
} from '@/types/api'

export const useAppAuthStore = defineStore('app-auth', () => {
  const status = ref<AppAuthStatus | null>(null)
  const loading = ref(false)
  let request: Promise<AppAuthStatus> | null = null

  const initialized = computed(() => status.value?.initialized === true)
  const authenticated = computed(() => status.value?.authenticated === true)

  async function load(force = false): Promise<AppAuthStatus> {
    if (status.value && !force) return status.value
    if (request) return request
    loading.value = true
    request = appAuthApi.status()
    try {
      status.value = await request
      return status.value
    } finally {
      request = null
      loading.value = false
    }
  }

  async function setup(payload: AppSetupRequest): Promise<AppAuthStatus> {
    loading.value = true
    try {
      status.value = await appAuthApi.setup(payload)
      return status.value
    } finally {
      loading.value = false
    }
  }

  async function login(payload: AppLoginRequest): Promise<AppAuthStatus> {
    loading.value = true
    try {
      status.value = await appAuthApi.login(payload)
      return status.value
    } finally {
      loading.value = false
    }
  }

  async function logout(): Promise<void> {
    loading.value = true
    try {
      status.value = await appAuthApi.logout()
    } finally {
      loading.value = false
    }
  }

  async function changePassword(payload: AppPasswordChangeRequest): Promise<void> {
    loading.value = true
    try {
      status.value = await appAuthApi.changePassword(payload)
    } finally {
      loading.value = false
    }
  }

  function expire(): void {
    status.value = status.value
      ? { ...status.value, authenticated: false, csrfToken: null, sessionExpiresAt: null }
      : null
  }

  return {
    status,
    loading,
    initialized,
    authenticated,
    load,
    setup,
    login,
    logout,
    changePassword,
    expire,
  }
})

File: backend/app/api/app_auth.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Header, Request, Response

from app.api.dependencies import get_app_auth_service
from app.schemas.app_auth import (
    AppAuthStatusRead,
    AppLoginRequest,
    AppPasswordChangeRequest,
    AppSetupRequest,
)
from app.services.app_auth import AppAuthService, AppPrincipal, CreatedAppSession

router = APIRouter(prefix="/app-auth", tags=["app-auth"])


def _token(request: Request, service: AppAuthService) -> str | None:
    return request.cookies.get(service.cookie_name)


def _set_session_cookie(
    response: Response,
    request: Request,
    service: AppAuthService,
    created: CreatedAppSession,
) -> None:
    response.set_cookie(
        key=service.cookie_name,
        value=created.token,
        max_age=service.settings.app_session_ttl_seconds,
        httponly=True,
        secure=service.settings.app_session_cookie_secure or request.url.scheme == "https",
        samesite="strict",
        path="/",
    )


async def require_app_session(
    request: Request,
    csrf_token: str | None = Header(default=None, alias="X-CSRF-Token"),
    service: AppAuthService = Depends(get_app_auth_service),
) -> AppPrincipal:
    principal = await service.authenticate(
        _token(request, service),
        csrf_token=csrf_token,
        require_csrf=request.method not in {"GET", "HEAD", "OPTIONS"},
    )
    if principal is None:
        raise RuntimeError("Required application session was not resolved")
    return principal


@router.get("/status", response_model=AppAuthStatusRead)
async def app_auth_status(
    request: Request,
    service: AppAuthService = Depends(get_app_auth_service),
) -> AppAuthStatusRead:
    return await service.status(_token(request, service))


@router.post("/setup", response_model=AppAuthStatusRead)
async def setup_app_admin(
    payload: AppSetupRequest,
    request: Request,
    response: Response,
    service: AppAuthService = Depends(get_app_auth_service),
) -> AppAuthStatusRead:
    created = await service.setup(payload)
    _set_session_cookie(response, request, service, created)
    return created.status


@router.post("/login", response_model=AppAuthStatusRead)
async def login_app_admin(
    payload: AppLoginRequest,
    request: Request,
    response: Response,
    service: AppAuthService = Depends(get_app_auth_service),
) -> AppAuthStatusRead:
    client_key = request.client.host if request.client else "unknown"
    created = await service.login(payload, client_key=client_key)
    _set_session_cookie(response, request, service, created)
    return created.status


@router.post("/logout", response_model=AppAuthStatusRead)
async def logout_app_admin(
    request: Request,
    response: Response,
    _: AppPrincipal = Depends(require_app_session),
    service: AppAuthService = Depends(get_app_auth_service),
) -> AppAuthStatusRead:
    await service.logout(_token(request, service))
    response.delete_cookie(service.cookie_name, path="/", httponly=True, samesite="strict")
    return AppAuthStatusRead(initialized=True, authenticated=False)


@router.put("/password", response_model=AppAuthStatusRead)
async def change_app_password(
    payload: AppPasswordChangeRequest,
    request: Request,
    response: Response,
    principal: AppPrincipal = Depends(require_app_session),
    service: AppAuthService = Depends(get_app_auth_service),
) -> AppAuthStatusRead:
    created = await service.change_password(principal, payload)
    _set_session_cookie(response, request, service, created)
    return created.status
